pkg/messaging/rabbitmq/pubsub.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package rabbitmq
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+const (
+	chansPrefix = "channels"
+	// SubjectAllChannels represents subject to subscribe for all the channels.
+	SubjectAllChannels = "channels.>"
+	exchangeName       = "mainflux-exchange"
+	exchangeKind       = "direct"
+)
+
+var (
+	ErrAlreadySubscribed = errors.New("already subscribed to topic")
+	ErrNotSubscribed     = errors.New("not subscribed")
+	ErrEmptyTopic        = errors.New("empty topic")
+	ErrEmptyID           = errors.New("empty id")
+)
+
+var _ messaging.PubSub = (*pubsub)(nil)
+
+// PubSub wraps messaging Publisher exposing
+// Close() method for RabbitMQ connection.
+type PubSub interface {
+	messaging.PubSub
+	Close()
+}
+
+type subscription struct {
+	cancel func() error
+}
+type pubsub struct {
+	publisher
+	logger        log.Logger
+	subscriptions map[string]map[string]subscription
+	mu            sync.Mutex
+}
+
+// NewPubSub returns RabbitMQ message publisher/subscriber.
+func NewPubSub(url, queue string, logger log.Logger) (PubSub, error) {
+	endpoint := fmt.Sprintf("amqp://%s", url)
+	conn, err := amqp.Dial(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	ch, err := conn.Channel()
+	if err != nil {
+		return nil, err
+	}
+	if err := ch.ExchangeDeclare(exchangeName, amqp.ExchangeDirect, true, false, false, false, nil); err != nil {
+		return nil, err
+	}
+	ret := &pubsub{
+		publisher: publisher{
+			conn: conn,
+			ch:   ch,
+		},
+		logger:        logger,
+		subscriptions: make(map[string]map[string]subscription),
+	}
+	return ret, nil
+}
+
+func (ps *pubsub) Subscribe(id, topic string, handler messaging.MessageHandler) error {
+	if id == "" {
+		return ErrEmptyID
+	}
+	if topic == "" {
+		return ErrEmptyTopic
+	}
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	// Check topic
+	s, ok := ps.subscriptions[topic]
+	switch ok {
+	case true:
+		// Check topic ID
+		if _, ok := s[id]; ok {
+			return ErrAlreadySubscribed
+		}
+	default:
+		s = make(map[string]subscription)
+		ps.subscriptions[topic] = s
+	}
+
+	_, err := ps.ch.QueueDeclare(topic, true, true, true, false, nil)
+	if err != nil {
+		return err
+	}
+	if err := ps.ch.QueueBind(topic, topic, exchangeName, false, nil); err != nil {
+		return err
+	}
+	msgs, err := ps.ch.Consume(topic, id, true, false, false, false, nil)
+	if err != nil {
+		return err
+	}
+	go ps.handle(msgs, handler)
+	s[id] = subscription{
+		cancel: handler.Cancel,
+	}
+
+	return nil
+}
+
+func (ps *pubsub) Unsubscribe(id, topic string) error {
+	defer ps.ch.Cancel(id, false)
```

Comment:
Should we really cancel unconditionally?

Reply:
From the docs, this is what it says about the Cancel method

> Only use this method to cleanly stop receiving deliveries from the server and cleanly shut down the consumer chan identified by this tag.  Using this method and waiting for the remaining messages to flush from the consumer chan will ensure all messages received on the network will be delivered to the receiver of your consumer chan.

I believe this is what we want to achieve

Reply:
We are, but we should first handle the internal logic, then try to cancel and handle the potential errors. What happens if we have the same consumer subscribed to multiple topics? Does it cancel them all (because it's the same consumer ID)?
For example, a nice solution would be to put this in the `cancel` method of `subscriber` when you subscribe, so that you don't even call it explicitly when unsubscribing. Something like:
```go
	s[id] = subscription{
		cancel: func() error {
			if err := ps.ch.Cancel(id, false); err != nil {
				return err
			}
			return handler.Cancel()
		},
	}
```